compiler/src/parsing/parser.mly

```diff
@@ -0,0 +1,706 @@
+  | SEMI opt_eols {}
+
+lbrack:
+  | LBRACK opt_eols %prec EOL {}
+
+lbrackrcaret:
+  | LBRACKRCARET opt_eols %prec EOL {}
+
+rbrack:
+  | opt_eols RBRACK {}
+
+lparen:
+  | LPAREN opt_eols %prec EOL {}
+
+rparen:
+  | opt_eols RPAREN {}
+
+lbrace:
+  | LBRACE opt_eols %prec EOL {}
+
+rbrace:
+  | opt_eols RBRACE {}
+
+lcaret:
+  | LCARET opt_eols {}
+
+rcaret:
+  | opt_eols RCARET {}
+
+comma:
+  | COMMA opt_eols %prec COMMA {}
+
+%inline colon:
+  | COLON opt_eols {}
+
+%inline dot:
+  | DOT opt_eols {}
+
+arrow:
+  | ARROW opt_eols {}
+
+thickarrow:
+  | opt_eols THICKARROW opt_eols {}
+
+equal:
+  | EQUAL opt_eols {}
+
+const:
+  // Rational literals are a special case of the division binop_expr.
+  | dash_op? NUMBER_INT { Const.number (PConstNumberInt (if Option.is_some $1 then "-" ^ $2 else $2)), $sloc }
+  | dash_op? NUMBER_FLOAT { Const.number (PConstNumberFloat (if Option.is_some $1 then "-" ^ $2 else $2)), $sloc }
+  | dash_op? INT32 { Const.int32 (if Option.is_some $1 then "-" ^ $2 else $2), $sloc }
+  | dash_op? INT64 { Const.int64 (if Option.is_some $1 then "-" ^ $2 else $2), $sloc }
+  | dash_op? FLOAT32 { Const.float32 (if Option.is_some $1 then "-" ^ $2 else $2), $sloc }
+  | dash_op? FLOAT64 { Const.float64 (if Option.is_some $1 then "-" ^ $2 else $2), $sloc }
+  | dash_op? WASMI32 { Const.wasmi32 (if Option.is_some $1 then "-" ^ $2 else $2), $sloc }
+  | dash_op? WASMI64 { Const.wasmi64 (if Option.is_some $1 then "-" ^ $2 else $2), $sloc }
+  | dash_op? WASMF32 { Const.wasmf32 (if Option.is_some $1 then "-" ^ $2 else $2), $sloc }
+  | dash_op? WASMF64 { Const.wasmf64 (if Option.is_some $1 then "-" ^ $2 else $2), $sloc }
+  | TRUE { Const.bool true, $loc }
+  | FALSE { Const.bool false, $loc }
+  | VOID { Const.void, $loc }
+  | STRING { Const.string $1, $loc }
+  | CHAR { Const.char $1, $loc }
+
+expr:
+  | stmt_expr { $1 }
+  // allow infix operators to cause a shift
+  | non_stmt_expr %prec _below_infix { $1 }
+
+non_binop_expr:
+  | lam_expr { $1 }
+  | non_assign_expr { $1 }
+  | assign_expr { $1 }
+
+non_stmt_expr:
+  | binop_expr { $1 }
+  | annotated_expr { $1 }
+
+%inline opt_annotation:
+  | %prec COLON { None }
+  | colon typ { Some $2 }
+
+annotated_expr:
+  | non_binop_expr opt_annotation { Option.fold ~none:$1 ~some:(fun ann -> Exp.constraint_ ~loc:(to_loc $loc) $1 ann) $2 }
+
+binop_expr:
+  | non_stmt_expr infix_op opt_eols non_stmt_expr { Exp.binop ~loc:(to_loc $loc) (mkid_expr $loc($2) [$2]) [$1; $4] }
+
+ellipsis_prefix(X):
+  | ELLIPSIS X {$2}
+
+pattern:
+  | pattern colon typ { Pat.constraint_ ~loc:(to_loc $loc) $1 $3 }
+  | UNDERSCORE { Pat.any ~loc:(to_loc $loc) () }
+  | const { let (pat, loc) = $1 in Pat.constant ~loc:(to_loc loc) pat }
+  // Allow rational numbers in patterns
+  | dash_op? NUMBER_INT slash_op dash_op? NUMBER_INT { Pat.constant ~loc:(to_loc $sloc) @@ Const.number (PConstNumberRational ((if Option.is_some $1 then "-" ^ $2 else $2), (if Option.is_some $4 then "-" ^ $5 else $5))) }
```

Comment:
Is there a reason you don't have a rule for rationals instead of using this in multiple places?

Reply:
I have no idea what I was seeing then